### PR TITLE
TFNetworkLayer: add DctLayer

### PR DESCRIPTION
**Pull request by vieting**
I added a `DctLayer` which wraps `tf.signal.dct`. So far, that's all it does, so no specific handling of axes is done. `tf.signal.dct` always computes the DCT over the last dimension.

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -4963,6 +4963,45 @@ def get_out_data_from_opts(cls, name, pool_size, strides=None, dilation_rate=1,
...
+  Wraps :func:`tf.signal.dct`. For further documentation on the input arguments, refer to
+  https://www.tensorflow.org/api_docs/python/tf/signal/dct
+  """
+
+  layer_class = "dct"
+  recurrent = True  # we should not shuffle in the time-dimension
+
+  def __init__(self, type=2, n=None, norm=None, **kwargs):
+    """
+    :param type: DCT type to perform
```
Comment (albertz) [inline TFNetworkLayer.py:4978]: Type missing. I guess `int`?

File: TFNetworkLayer.py [hunk 2]
```diff
@@ -4963,6 +4963,45 @@ def get_out_data_from_opts(cls, name, pool_size, strides=None, dilation_rate=1,
...
+    :param n: length of the transform
```
Comment (albertz) [inline TFNetworkLayer.py:4979]: Type missing. I guess `int`?

File: TFNetworkLayer.py [hunk 3]
```diff
@@ -4963,6 +4963,45 @@ def get_out_data_from_opts(cls, name, pool_size, strides=None, dilation_rate=1,
...
+    :param norm: normalization to apply
```
Comment (albertz) [inline TFNetworkLayer.py:4980]: Type missing? I have no idea, maybe `str|None`?

File: TFNetworkLayer.py [hunk 4]
```diff
@@ -4963,6 +4963,45 @@ def get_out_data_from_opts(cls, name, pool_size, strides=None, dilation_rate=1,
...
+    """
+    assert n is None, 'Not implemented yet for n other than None'
+    super(DctLayer, self).__init__(**kwargs)
+    input_data = self.input_data.copy_as_batch_spatial_major()
+    x = input_data.placeholder
+    y = tf.spectral.dct(x, type=type, n=n, norm=norm)
```
Comment (albertz) [inline TFNetworkLayer.py:4986]: Better use `tf.signal.dct`. I think `tf.spectral` is deprecated since a long time.

Comment (vieting) [reply TFNetworkLayer.py:4986]: `tf.signal` only works with tf versions >1.12. I am mostly using 1.8, that's why I used `tf.spectral`. Should we add some workaround for older versions?

Comment (albertz) [reply TFNetworkLayer.py:4986]: But I think this code does not work on TF 2 then. And I would say that supporting a recent TF version has a higher priority than a very outdated TF version. Also see the contributing guidelines about this. There is the rule that new code must work with a recent TF version.

You could maybe use `TFCompat.v1.spectral.dct`.
Or you could simply update your TF version, which I anyway would recommend.

Comment (vieting) [reply TFNetworkLayer.py:4986]: I agree of course to the priorities.

> You could maybe use `TFCompat.v1.spectral.dct`

Do you mean as kind of a fallback if `tf.signal` is not working?

Comment (albertz) [reply TFNetworkLayer.py:4986]: The code should also be simple, so avoid having two code paths there.

Comment (vieting) [reply TFNetworkLayer.py:4986]: I'll add the compatibility version then for now.

**Pull request merged**